### PR TITLE
fix: type issues in verifier's code

### LINKED ISSUE #1268: type issues in verifier's code
Recent changes increased the number of type issues reported by `pyright` on the verifier's code. The following two things are the most worrying to me:

- the outstanding issue of not being able to initialize IMA policy (that would make the type-checker happy) without breaking the functionality of the code
- plenty of type issues around the new caching mechanism

I can probably do something about the latter point, but about the former one I really need help from someone with a better understanding of the code.

### PR DESCRIPTION
**Pull request by mdrocco**
Fixes #1268

### CODE REVIEW

File: keylime/cloud_verifier_tornado.py [hunk 1]
```diff
@@ -129,24 +129,24 @@ def _from_db_obj(agent_db_obj: VerfierMain) -> Dict[str, str]:
 
 def verifier_read_policy_from_cache(stored_agent: VerfierMain) -> str:
     if stored_agent.agent_id not in GLOBAL_POLICY_CACHE:
-        GLOBAL_POLICY_CACHE[stored_agent.agent_id] = {}
+        GLOBAL_POLICY_CACHE[str(stored_agent.agent_id)] = {}
```
Comment (aplanas) [inline keylime/cloud_verifier_tornado.py:132]: I see ... the database should return here a string. Is it the complain that it is detecting a Column?

**Comment by mdrocco** [comment]
> @mdrocco against which pyright version are you testing, because if it is the current released one we should have seen those errors the CI or not?

Not yet, verifier's code is ignored in the current `pyproject.toml`.

**Comment by THS-on** [comment]
> Not yet, verifier's code is ignored in the current pyproject.toml.

Ah you are right!

File: keylime/cloud_verifier_tornado.py (hunk 1 above)
Comment (aplanas) [reply keylime/cloud_verifier_tornado.py:132]: no, I think that this makes sense

**Comment by mdrocco** [comment]
After rebasing on `master` and fixing some details, farm tests are now passing. However, type checking is failing on the issue about asynchronous callbacks that I fixed in the past, although I do not see that issue reported in my local setup (with the same pyright version). I suspect pyright might be misbehaving, but please could someone more knowledgeable than me take a look? @stefanberger @aplanas

**Comment by stefanberger** [comment]
On my machine with python 3.9.14 your branch passes all tests just fine.
On my machine with python 3.6.8 I get the following errors reported by latest pyright:

```
/home/stefanb/keylime/keylime/agentstates.py
  /home/stefanb/keylime/keylime/agentstates.py:31:16 - error: Expected no arguments to "TPMClockInfo" constructor (reportGeneralTypeIssues)
  /home/stefanb/keylime/keylime/agentstates.py:110:30 - error: Expected no arguments to "TPMClockInfo" constructor (reportGeneralTypeIssues)
/home/stefanb/keylime/keylime/cloud_verifier_tornado.py
  /home/stefanb/keylime/keylime/cloud_verifier_tornado.py:20:32 - error: "NoResultFound" is unknown import symbol (reportGeneralTypeIssues)
  /home/stefanb/keylime/keylime/cloud_verifier_tornado.py:1318:80 - error: Argument of type "() -> Coroutine[Any, Any, None]" cannot be assigned to parameter "callback" of type "(...) -> None" in function "call_later"
    Type "() -> Coroutine[Any, Any, None]" cannot be assigned to type "(...) -> None"
      Function return type "Coroutine[Any, Any, None]" is incompatible with type "None"
        Type cannot be assigned to type "None" (reportGeneralTypeIssues)
  /home/stefanb/keylime/keylime/cloud_verifier_tornado.py:1354:72 - error: Argument of type "() -> Coroutine[Any, Any, None]" cannot be assigned to parameter "callback" of type "(...) -> None" in function "call_later"
    Type "() -> Coroutine[Any, Any, None]" cannot be assigned to type "(...) -> None"
      Function return type "Coroutine[Any, Any, None]" is incompatible with type "None"
        Type cannot be assigned to type "None" (reportGeneralTypeIssues)
  /home/stefanb/keylime/keylime/cloud_verifier_tornado.py:1382:72 - error: Argument of type "() -> Coroutine[Any, Any, None]" cannot be assigned to parameter "callback" of type "(...) -> None" in function "call_later"
    Type "() -> Coroutine[Any, Any, None]" cannot be assigned to type "(...) -> None"
      Function return type "Coroutine[Any, Any, None]" is incompatible with type "None"
        Type cannot be assigned to type "None" (reportGeneralTypeIssues)
/home/stefanb/keylime/keylime/registrar_common.py
  /home/stefanb/keylime/keylime/registrar_common.py:14:32 - error: "NoResultFound" is unknown import symbol (reportGeneralTypeIssues)
/home/stefanb/keylime/keylime/tenant.py
  /home/stefanb/keylime/keylime/tenant.py:559:23 - error: Object of type "None" is not subscriptable (reportOptionalSubscript)
  /home/stefanb/keylime/keylime/tenant.py:560:46 - error: "get" is not a known member of "None" (reportOptionalMemberAccess)
/home/stefanb/keylime/keylime/da/attest.py
  /home/stefanb/keylime/keylime/da/attest.py:50:33 - error: Cannot access member "fromisoformat" for type "Type[datetime]"
    Member "fromisoformat" is unknown (reportGeneralTypeIssues)
  /home/stefanb/keylime/keylime/da/attest.py:57:35 - error: Cannot access member "fromisoformat" for type "Type[datetime]"
    Member "fromisoformat" is unknown (reportGeneralTypeIssues)
/home/stefanb/keylime/keylime/db/keylime_db.py
  /home/stefanb/keylime/keylime/db/keylime_db.py:90:16 - error: Expression of type "MockConnection" cannot be assigned to return type "Engine"
    "MockConnection" is incompatible with "Engine" (reportGeneralTypeIssues)
/home/stefanb/keylime/keylime/ima/ima_dm.py
  /home/stefanb/keylime/keylime/ima/ima_dm.py:362:36 - error: Expected no arguments to "DeviceState" constructor (reportGeneralTypeIssues)
  /home/stefanb/keylime/keylime/ima/ima_dm.py:827:16 - error: Expected no arguments to "VerityAttributes" constructor (reportGeneralTypeIssues)
  /home/stefanb/keylime/keylime/ima/ima_dm.py:831:16 - error: Expected no arguments to "CacheAttributes" constructor (reportGeneralTypeIssues)
  /home/stefanb/keylime/keylime/ima/ima_dm.py:835:16 - error: Expected no arguments to "CryptAttributes" constructor (reportGeneralTypeIssues)
  /home/stefanb/keylime/keylime/ima/ima_dm.py:839:16 - error: Expected no arguments to "IntegrityAttributes" constructor (reportGeneralTypeIssues)
  /home/stefanb/keylime/keylime/ima/ima_dm.py:843:16 - error: Expected no arguments to "LinearAttributes" constructor (reportGeneralTypeIssues)
  /home/stefanb/keylime/keylime/ima/ima_dm.py:847:16 - error: Expected no arguments to "MirrorAttributes" constructor (reportGeneralTypeIssues)
  /home/stefanb/keylime/keylime/ima/ima_dm.py:851:16 - error: Expected no arguments to "SnapshotAttributes" constructor (reportGeneralTypeIssues)
  /home/stefanb/keylime/keylime/ima/ima_dm.py:855:27 - error: Expected no arguments to "Target" constructor (reportGeneralTypeIssues)
  /home/stefanb/keylime/keylime/ima/ima_dm.py:866:36 - error: Expected no arguments to "DeviceMetaData" constructor (reportGeneralTypeIssues)
  /home/stefanb/keylime/keylime/ima/ima_dm.py:870:16 - error: Expected no arguments to "LoadEvent" constructor (reportGeneralTypeIssues)
  /home/stefanb/keylime/keylime/ima/ima_dm.py:875:16 - error: Expected no arguments to "ResumeEvent" constructor (reportGeneralTypeIssues)
  /home/stefanb/keylime/keylime/ima/ima_dm.py:880:16 - error: Expected no arguments to "RemoveEvent" constructor (reportGeneralTypeIssues)
  /home/stefanb/keylime/keylime/ima/ima_dm.py:884:16 - error: Expected no arguments to "RenameEvent" constructor (reportGeneralTypeIssues)
  /home/stefanb/keylime/keylime/ima/ima_dm.py:889:16 - error: Expected no arguments to "ClearEvent" constructor (reportGeneralTypeIssues)
  /home/stefanb/keylime/keylime/ima/ima_dm.py:893:16 - error: Expected no arguments to "UpdateEvent" constructor (reportGeneralTypeIssues)
  /home/stefanb/keylime/keylime/ima/ima_dm.py:928:39 - error: Expected no arguments to "DeviceMetaDataMinimal" constructor (reportGeneralTypeIssues)
  /home/stefanb/keylime/keylime/ima/ima_dm.py:945:38 - error: Expected no arguments to "MirrorDevice" constructor (reportGeneralTypeIssues)
```

That's a lot more issue than what the test machine is reporting with python 3.9 and my guess it comes from how pyright behaves with the underlying python version. Maybe we need a minimum python version.

**Comment by mdrocco** [comment]
@stefanberger I am not clear of what you are proposing by the "minimum python version", do you mean it as a general requirement for Keylime or limited to type (and code quality) checking?

**Comment by stefanberger** [comment]
I mean that we would require a minimum python version for style checking. because it seemed odd that no errors are reported on more recent version of python while pyright on older versions reports lots of errors.

**Comment by mpeters** [comment]
I'm fine with requiring a specific version for type checking. It seems like it would be hairy to get it right for all of them since python itself sees differences.

**Comment by stefanberger** [comment]
> I'm fine with requiring a specific version for type checking. It seems like it would be hairy to get it right for all of them since python itself sees differences.

Since the problems are not widespread maybe this would suffice:

```
                pending = tornado.ioloop.IOLoop.current().call_later(
                    interval, invoke_get_quote, agent, runtime_policy, False, timeout=timeout  # type: ignore  ; due to python <3.9
                )
```

I am not sure about the `<3.9` comment though, it could be `<3.8` also.

**Comment by aplanas** [comment]
> `# type: ignore  ; due to python <3.9`

I agree. I would agree also on pinning the most up to date Python (3.11) for type checking tests in the CI, but running the rest of the tests in the different supported Python version (>=3.6). There is no gain in running pyright in 3.6..=3.11, and fix failures old Python

**Comment by aplanas** [comment]
Safe to ignore: add `# pylint: disable=too-many-ancestors` in the line 14 of setup.py.

To learn more check this: https://pylint.pycqa.org/en/latest/user_guide/messages/refactor/too-many-ancestors.html

But the tl;dr is that keylime_build is inheriting from build_py and pylint found (not sure where) other classes inheriting from the same base class, and decides that 8 is too much and 7 is OK

**Comment by stefanberger** [comment]
pyright issue: https://github.com/microsoft/pyright/issues/4574

**Comment by stefanberger** [comment]
Now the issue is here: https://github.com/RobertCraigie/pyright-python/issues/145

File: pyproject.toml [hunk 2]
```diff
@@ -28,7 +28,6 @@ ignore = [
     "keylime/da/examples/redis.py",
     "keylime/da/examples/file.py",
     "keylime/da/record.py",
-    "keylime/cloud_verifier_tornado.py",
```
Comment (stefanberger) [inline pyproject.toml:31]: My system spits out tons of errors now::

```
Assuming Python version 3.9
Assuming Python platform Linux
Auto-excluding **/node_modules
Auto-excluding **/__pycache__
Auto-excluding **/.*
stubPath /home/stefanb/keylime/typings is not a valid directory.
Searching for source files
Found 148 source files
pyright 1.1.292
/home/stefanb/keylime/keylime/cloud_verifier_tornado.py
  /home/stefanb/keylime/keylime/cloud_verifier_tornado.py:20:32 - error: "NoResultFound" is not exported from module "sqlalchemy.orm.exc"
    Import from "sqlalchemy.exc" instead (reportPrivateImportUsage)
  /home/stefanb/keylime/keylime/cloud_verifier_tornado.py:167:26 - error: "boottime" is not a known member of "None" (reportOptionalMemberAccess)
  /home/stefanb/keylime/keylime/cloud_verifier_tornado.py:168:26 - error: "next_ima_ml_entry" is not a known member of "None" (reportOptionalMemberAccess)
  /home/stefanb/keylime/keylime/cloud_verifier_tornado.py:170:26 - error: "ima_pcrs" is not a known member of "None" (reportOptionalMemberAccess)
  /home/stefanb/keylime/keylime/cloud_verifier_tornado.py:173:26 - error: "learned_ima_keyrings" is not a known member of "None" (reportOptionalMemberAccess)
  /home/stefanb/keylime/keylime/cloud_verifier_tornado.py:327:29 - error: Argument of type "Column[str]" cannot be assigned to parameter "attrs" of type "_AttrType" in function "load_only"
    Type "Column[str]" cannot be assigned to type "_AttrType"
      "Column[str]" is incompatible with "str"
      "Column[str]" is incompatible with "QueryableAttribute[Any]" (reportGeneralTypeIssues)
  /home/stefanb/keylime/keylime/cloud_verifier_tornado.py:327:57 - error: Argument of type "Column[int]" cannot be assigned to parameter "attrs" of type "_AttrType" in function "load_only"
    Type "Column[int]" cannot be assigned to type "_AttrType"
      "Column[int]" is incompatible with "str"
      "Column[int]" is incompatible with "QueryableAttribute[Any]" (reportGeneralTypeIssues)
  /home/stefanb/keylime/keylime/cloud_verifier_tornado.py:351:33 - error: Argument of type "Column[str]" cannot be assigned to parameter "attrs" of type "_AttrType" in function "load_only"
    Type "Column[str]" cannot be assigned to type "_AttrType"
      "Column[str]" is incompatible with "str"
      "Column[str]" is incompatible with "QueryableAttribute[Any]" (reportGeneralTypeIssues)
  /home/stefanb/keylime/keylime/cloud_verifier_tornado.py:351:61 - error: Argument of type "Column[int]" cannot be assigned to parameter "attrs" of type "_AttrType" in function "load_only"
    Type "Column[int]" cannot be assigned to type "_AttrType"
      "Column[int]" is incompatible with "str"
      "Column[int]" is incompatible with "QueryableAttribute[Any]" (reportGeneralTypeIssues)
  /home/stefanb/keylime/keylime/cloud_verifier_tornado.py:362:33 - error: Argument of type "Column[str]" cannot be assigned to parameter "attrs" of type "_AttrType" in function "load_only"
    Type "Column[str]" cannot be assigned to type "_AttrType"
      "Column[str]" is incompatible with "str"
      "Column[str]" is incompatible with "QueryableAttribute[Any]" (reportGeneralTypeIssues)
  /home/stefanb/keylime/keylime/cloud_verifier_tornado.py:362:61 - error: Argument of type "Column[int]" cannot be assigned to parameter "attrs" of type "_AttrType" in function "load_only"
    Type "Column[int]" cannot be assigned to type "_AttrType"
      "Column[int]" is incompatible with "str"
      "Column[int]" is incompatible with "QueryableAttribute[Any]" (reportGeneralTypeIssues)
  /home/stefanb/keylime/keylime/cloud_verifier_tornado.py:459:30 - error: "operational_state" is not a known member of "None" (reportOptionalMemberAccess)
  /home/stefanb/keylime/keylime/cloud_verifier_tornado.py:1235:71 - error: Argument of type "Column[str]" cannot be assigned to parameter "attrs" of type "_AttrType" in function "load_only"
    Type "Column[str]" cannot be assigned to type "_AttrType"
      "Column[str]" is incompatible with "str"
      "Column[str]" is incompatible with "QueryableAttribute[Any]" (reportGeneralTypeIssues)
  /home/stefanb/keylime/keylime/cloud_verifier_tornado.py:1012:89 - error: Argument of type "Dict[str, Any]" cannot be assigned to parameter "values" of type "Dict[_DMLColumnArgument, Any]" in function "update"
    "Dict[str, Any]" is incompatible with "Dict[_DMLColumnArgument, Any]"
      TypeVar "_KT@dict" is invariant
        Type "str" cannot be assigned to type "_DMLColumnArgument"
          "str" is incompatible with "ColumnClause[Any]"
          "str" is incompatible with protocol "_HasClauseElement" (reportGeneralTypeIssues)
  /home/stefanb/keylime/keylime/cloud_verifier_tornado.py:1193:44 - error: Argument of type "Dict[str, Any]" cannot be assigned to parameter "tosend" of type "bytes" in function "notify_webhook"
    "Dict[str, Any]" is incompatible with "bytes" (reportGeneralTypeIssues)
  /home/stefanb/keylime/keylime/cloud_verifier_tornado.py:1290:93 - error: Argument of type "Dict[str, Any]" cannot be assigned to parameter "values" of type "Dict[_DMLColumnArgument, Any]" in function "update"
    "Dict[str, Any]" is incompatible with "Dict[_DMLColumnArgument, Any]"
      TypeVar "_KT@dict" is invariant
        Type "str" cannot be assigned to type "_DMLColumnArgument"
          "str" is incompatible with "ColumnClause[Any]"
          "str" is incompatible with protocol "_HasClauseElement" (reportGeneralTypeIssues)
  /home/stefanb/keylime/keylime/cloud_verifier_tornado.py:1300:88 - error: Argument of type "dict[str, Any]" cannot be assigned to parameter "values" of type "Dict[_DMLColumnArgument, Any]" in function "update"
    "dict[str, Any]" is incompatible with "Dict[_DMLColumnArgument, Any]"
      TypeVar "_KT@dict" is invariant
        Type "str" cannot be assigned to type "_DMLColumnArgument"
          "str" is incompatible with "ColumnClause[Any]"
          "str" is incompatible with protocol "_HasClauseElement" (reportGeneralTypeIssues)
  /home/stefanb/keylime/keylime/cloud_verifier_tornado.py:1428:19 - error: Cannot assign member "verifier_ip" for type "VerfierMain"
    Expression of type "str" cannot be assigned to member "verifier_ip" of class "VerfierMain"
      "str" is incompatible with "Column[str]" (reportGeneralTypeIssues)
  /home/stefanb/keylime/keylime/cloud_verifier_tornado.py:1443:21 - error: Argument of type "Column[str]" cannot be assigned to parameter "agent_id" of type "str" in function "add"
    "Column[str]" is incompatible with "str" (reportGeneralTypeIssues)
  /home/stefanb/keylime/keylime/cloud_verifier_tornado.py:1443:37 - error: Argument of type "Column[int]" cannot be assigned to parameter "boottime" of type "int" in function "add"
    "Column[int]" is incompatible with "int" (reportGeneralTypeIssues)
  /home/stefanb/keylime/keylime/cloud_verifier_tornado.py:1443:68 - error: Argument of type "Column[int]" cannot be assigned to parameter "next_ima_ml_entry" of type "int" in function "add"
    "Column[int]" is incompatible with "int" (reportGeneralTypeIssues)
  /home/stefanb/keylime/keylime/cloud_verifier_tornado.py:1443:93 - error: Argument of type "Column[object]" cannot be assigned to parameter "learned_ima_keyrings" of type "Dict[str, str]" in function "add"
    "Column[object]" is incompatible with "Dict[str, str]" (reportGeneralTypeIssues)
  /home/stefanb/keylime/keylime/cloud_verifier_tornado.py:1480:21 - error: Cannot assign member "operational_state" for type "VerfierMain"
    Expression of type "Literal[1]" cannot be assigned to member "operational_state" of class "VerfierMain"
      "Literal[1]" is incompatible with "Column[int]" (reportGeneralTypeIssues)
/home/stefanb/keylime/keylime/registrar_common.py
  /home/stefanb/keylime/keylime/registrar_common.py:14:32 - error: "NoResultFound" is not exported from module "sqlalchemy.orm.exc"
    Import from "sqlalchemy.exc" instead (reportPrivateImportUsage)
24 errors, 0 warnings, 0 informations
```

**Review by stefanberger — CHANGES_REQUESTED** [review]

**Comment by stefanberger** [comment]
The pyright/nodeenv issue may be intermittent. Maybe give this branch another push...

**Review by maugustosilva — APPROVED** [review]

**Comment by stefanberger** [comment]
You could squash the two topmost patches since they cancel each other out. Otherwise LGTM.

**Comment by maugustosilva** [comment]
All tests are passing now, LGTM

**Comment by mdrocco** [comment]
@stefanberger why don't we just "squash and merge" the whole thing?

**Comment by mdrocco** [comment]
I rebased and squashed

**Comment by mdrocco** [comment]
thanks @kkaarreell for pointing this out.

**Pull request merged**